lsp/lsp.cpp

```diff
@@ -73,265 +73,281 @@ static const char *MethodNotFound       = "-32601";
-    Lexer lex(runtime.heap, filePath.c_str());
-    parse_top(*top, lex);
-  }
-
-  JAST diagnosticsArray(JSON_ARRAY);    
-  for (Diagnostic diagnostic: diagnostics) {
-    diagnosticsArray.children.emplace_back("", createDiagnostic(diagnostic)); // add .add for JSON_OBJECT to JSON_ARRAY
-  }
-  JAST message = createDiagnosticMessage();
-  JAST &params = message.add("params", JSON_OBJECT);
-  params.add("uri", fileUri.c_str());
-  params.children.emplace_back("diagnostics", diagnosticsArray);
-  diagnostics.clear();
-  sendMessage(message);  
-}
+      JAST &end = range.add("end", JSON_OBJECT);
+      end.add("line", std::max(0, diagnostic.getLocation().end.row));
+      end.add("character", std::max(0, diagnostic.getLocation().end.column)); // It can be -1
 
-void didOpen(JAST receivedMessage) {
-  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
-  diagnoseFile(fileUri);
-}
+      return range;
+    }
 
-void didChange(JAST receivedMessage) {
-  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
-  std::string fileContent = receivedMessage.get("params").get("contentChanges").children.back().second.get("text").value;
-  changedFiles[fileUri] = fileContent;
-  diagnoseFile(fileUri);
-}
+    JAST createDiagnostic(Diagnostic diagnostic) {
+      JAST diagnosticJSON(JSON_OBJECT);
+      
+      diagnosticJSON.children.emplace_back("range", createDiagnosticRange(diagnostic));
+      diagnosticJSON.add("severity", diagnostic.getSeverity());
+      diagnosticJSON.add("source", "wake");
+      JAST range = createDiagnosticRange(diagnostic);
 
-void didSave(JAST receivedMessage) {
-  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
-  changedFiles.erase(fileUri);
-  diagnoseFile(fileUri);
-}
+      diagnosticJSON.add("message", diagnostic.getMessage());
 
-void didClose(JAST receivedMessage) {
-  std::string fileUri = receivedMessage.get("params").get("textDocument").get("uri").value;
-  changedFiles.erase(fileUri);
-}
+      return diagnosticJSON;
+    }
 
-void didChangeWatchedFiles(JAST receivedMessage) {
-  JAST files = receivedMessage.get("params").get("changes");
-  for (auto child: files.children) {
-    std::string fileUri = child.second.get("uri").value;
-    changedFiles.erase(fileUri);
-    diagnoseFile(fileUri);
-  }       
-}
+    JAST createDiagnosticMessage() {
+      JAST message = createMessage();
+      message.add("method", "textDocument/publishDiagnostics");
+      return message;
+    }
 
-void shutdown(JAST receivedMessage) {
-  JAST message = createResponseMessage(receivedMessage);
-  message.add("result", JSON_NULLVAL);
-  isShutDown = true;
-  sendMessage(message);
-}
+    void diagnoseFile(std::string fileUri) {
+      std::string filePath = fileUri.substr(rootUri.length() + 1, std::string::npos);
+      std::unique_ptr<Top> top(new Top);
+
+      auto fileChangesPointer = changedFiles.find(fileUri);
```

Comment:
I'd suggest you create a new/local LSPReporter here, and then assign the global pointer to it. In the future, we'd like to pass the reporter TO the lexer, so there is no need for it to exist for longer than this call.

Reply:
Done.